mammoth/bin/train.py

```diff
@@ -201,92 +201,94 @@ def train(opts):
         node_rank = opts.node_rank
-
-        queues = []
-        semaphores = []
-        mp = torch.multiprocessing.get_context('spawn')
-        logger.info("world_size = {}, queue_size = {}".format(opts.world_size, opts.queue_size))
-        # Create a thread to listen for errors in the child processes.
-        error_queue = mp.SimpleQueue()
-        error_handler = ErrorHandler(error_queue)
-        # Train with multiprocessing.
-        procs = []
-        producers = []
-
-        for local_rank in range(world_context.gpus_per_node):
+        n_local_ranks = world_context.gpus_per_node
+    else:
+        n_local_ranks = 1
+
+    queues = []
+    semaphores = []
+    mp = torch.multiprocessing.get_context('spawn')
+    logger.info("world_size = {}, queue_size = {}".format(opts.world_size, opts.queue_size))
+    # Create a thread to listen for errors in the child processes.
+    error_queue = mp.SimpleQueue()
+    error_handler = ErrorHandler(error_queue)
+    # Train with multiprocessing.
+    procs = []
+    producers = []
+
+    for local_rank in range(n_local_ranks):
+        if world_context.context == DeviceContextEnum.MULTI_GPU:
             device_context: DeviceContext = world_context.global_to_local(
                 node_rank=node_rank,
                 local_rank=local_rank,
             )
-            # This task_queue_manager will only yield the items that are active on this gpu
-            task_queue_manager = global_task_queue_manager.global_to_local(
-                node_rank=node_rank,
-                local_rank=local_rank,
-                opts=opts
-            )
-
             # store rank in env (FIXME: is this obsolete?)
             current_env["RANK"] = str(device_context.global_rank)
             current_env["LOCAL_RANK"] = str(device_context.local_rank)
-
-            q = mp.Queue(opts.queue_size)
-            semaphore = mp.Semaphore(opts.queue_size)
-            queues.append(q)
-            semaphores.append(semaphore)
-            procs.append(
-                mp.Process(
-                    target=consumer,
-                    args=(
-                        train_process,
-                        opts,
-                        device_context,
-                        error_queue,
-                        q,
-                        semaphore,
-                        task_queue_manager,
-                        checkpoint),
-                    daemon=True,
-                ))
-            procs[local_rank].start()
-            logger.info(" Starting process pid: %d  " % procs[local_rank].pid)
-            error_handler.add_child(procs[local_rank].pid)
-
-            # Get the iterator to generate from
-            train_iter = DynamicDatasetIter.from_opts(
-                task_queue_manager=task_queue_manager,
-                transforms_cls=transforms_cls,
-                vocabs_dict=vocabs_dict,
-                opts=opts,
-                is_train=True,
+        else:
```

Comment:
maybe add a comment to stress this is a non-distributed context conditional branch